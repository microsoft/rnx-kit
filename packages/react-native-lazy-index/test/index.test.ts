import type { BabelFileResult } from "@babel/core";

describe("react-native-lazy-index", () => {
  const babel = require("@babel/core");
  const { spawnSync } = require("child_process");
  const path = require("path");

  const currentWorkingDir = process.cwd();

  /**
   * Generates a sequence from RegEx matches.
   */
  function* generateSequence(
    str: string,
    regex: RegExp
  ): Generator<string, void> {
    let m = regex.exec(str);
    while (m) {
      yield m[1];
      m = regex.exec(str);
    }
  }

  /**
   * Tests the specified fixture.
   */
  function transformFixture(fixture: string): BabelFileResult | null {
    const workingDir = path.join(__dirname, "__fixtures__", fixture);
    process.chdir(workingDir);
    return babel.transformFileSync("../../../src/index.js", {
      cwd: workingDir,
      filename: `${fixture}.js`,
      plugins: ["codegen"],
    });
  }

  afterEach(() => process.chdir(currentWorkingDir));

  test("wraps AppRegistry.registerComponent calls", () => {
    const result = transformFixture("AppRegistry");
    expect(result).toBeTruthy();
    expect(result?.code).toMatchSnapshot();
  });

  test("wraps BatchedBridge.registerCallableModule calls", () => {
    const result = transformFixture("BatchedBridge");
    expect(result).toBeTruthy();
    expect(result?.code).toMatchSnapshot();
  });

  test("wraps registered components", () => {
    const result = transformFixture("MyAwesomeApp");
    expect(result).toBeTruthy();
    expect(result?.code).toMatchSnapshot();
  });

  test("wraps registered components using declared entry points", () => {
    const result = transformFixture("MyOtherAwesomeApp");
    expect(result).toBeTruthy();
    expect(result?.code).toMatchSnapshot();
  });

  test("packs only necessary files", () => {
    const files = Array.from(
      generateSequence(
        spawnSync("npm", ["pack", "--dry-run"]).output.toString(),
        /[.\d]+k?B\s+([^\s]*)/g
      )
    );
    expect(files.sort()).toEqual([
      "LICENSE",
      "README.md",
      "package.json",
      "src/experiences.js",
      "src/generateLazyIndex.js",
      "src/index.js",
      "src/module.js",
    ]);
  });
});
