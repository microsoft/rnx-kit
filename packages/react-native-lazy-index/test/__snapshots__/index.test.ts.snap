// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-native-lazy-index wraps AppRegistry.registerComponent calls 1`] = `
"// @ts-nocheck
const {
  AppRegistry
} = require(\\"react-native\\");

AppRegistry.registerComponent(\\"Component-01\\", () => {
  // Source: component-01.js
  require(\\"./component-01\\");

  return AppRegistry.getRunnable(\\"Component-01\\").componentProvider();
});
AppRegistry.registerComponent(\\"Component-02\\", () => {
  // Source: component-02.js
  require(\\"./component-02\\");

  return AppRegistry.getRunnable(\\"Component-02\\").componentProvider();
});"
`;

exports[`react-native-lazy-index wraps BatchedBridge.registerCallableModule calls 1`] = `
"// @ts-nocheck
const BatchedBridge = require(\\"react-native/Libraries/BatchedBridge/BatchedBridge\\");

BatchedBridge.registerLazyCallableModule(\\"Module-01\\", () => {
  // Source: module-01.js
  require(\\"./module-01\\");

  return BatchedBridge.getCallableModule(\\"Module-01\\");
});
BatchedBridge.registerLazyCallableModule(\\"Module-02\\", () => {
  // Source: module-02.js
  require(\\"./module-02\\");

  return BatchedBridge.getCallableModule(\\"Module-02\\");
});"
`;

exports[`react-native-lazy-index wraps registered components 1`] = `
"// @ts-nocheck
const {
  AppRegistry
} = require(\\"react-native\\");

const BatchedBridge = require(\\"react-native/Libraries/BatchedBridge/BatchedBridge\\");

BatchedBridge.registerLazyCallableModule(\\"YetAnotherFeature\\", () => {
  // Source: node_modules/@awesome-app/yet-another-feature/index.js
  require(\\"@awesome-app/yet-another-feature\\");

  return BatchedBridge.getCallableModule(\\"YetAnotherFeature\\");
});
BatchedBridge.registerLazyCallableModule(\\"YetAnotherFeatureLazy\\", () => {
  // Source: node_modules/@awesome-app/yet-another-feature/index.js
  require(\\"@awesome-app/yet-another-feature\\");

  return BatchedBridge.getCallableModule(\\"YetAnotherFeatureLazy\\");
});
BatchedBridge.registerLazyCallableModule(\\"AnotherFeature\\", () => {
  // Source: node_modules/@awesome-app/another-feature/index.js
  require(\\"@awesome-app/another-feature\\");

  return BatchedBridge.getCallableModule(\\"AnotherFeature\\");
});
AppRegistry.registerComponent(\\"SomeFeature\\", () => {
  // Source: node_modules/@awesome-app/some-feature/index.js
  require(\\"@awesome-app/some-feature\\");

  return AppRegistry.getRunnable(\\"SomeFeature\\").componentProvider();
});
AppRegistry.registerComponent(\\"FinalFeature\\", () => {
  // Source: node_modules/@awesome-app/final-feature/index.js
  require(\\"@awesome-app/final-feature\\");

  return AppRegistry.getRunnable(\\"FinalFeature\\").componentProvider();
});"
`;

exports[`react-native-lazy-index wraps registered components using declared entry points 1`] = `
"// @ts-nocheck
const {
  AppRegistry
} = require(\\"react-native\\");

const BatchedBridge = require(\\"react-native/Libraries/BatchedBridge/BatchedBridge\\");

BatchedBridge.registerLazyCallableModule(\\"YetAnotherFeature\\", () => {
  // Source: package.json
  require(\\"@awesome-app/yet-another-feature\\");

  return BatchedBridge.getCallableModule(\\"YetAnotherFeature\\");
});
BatchedBridge.registerLazyCallableModule(\\"YetAnotherFeatureLazy\\", () => {
  // Source: package.json
  require(\\"@awesome-app/yet-another-feature\\");

  return BatchedBridge.getCallableModule(\\"YetAnotherFeatureLazy\\");
});
BatchedBridge.registerLazyCallableModule(\\"AnotherFeature\\", () => {
  // Source: package.json
  require(\\"@awesome-app/another-feature\\");

  return BatchedBridge.getCallableModule(\\"AnotherFeature\\");
});
AppRegistry.registerComponent(\\"SomeFeature\\", () => {
  // Source: package.json
  require(\\"@awesome-app/some-feature\\");

  return AppRegistry.getRunnable(\\"SomeFeature\\").componentProvider();
});
AppRegistry.registerComponent(\\"FinalFeature\\", () => {
  // Source: package.json
  require(\\"@awesome-app/final-feature\\");

  return AppRegistry.getRunnable(\\"FinalFeature\\").componentProvider();
});"
`;
